src/zone/worker.cpp

```diff
@@ -132,15 +140,23 @@ void Worker::WorkerThreadFunc(const settings::ZoneSettings& settings) {
 
         {
             std::unique_lock<std::mutex> lock(_impl->queueLock);
-            if (_impl->tasks.empty()) {
+            if (_impl->tasks.empty() && _impl->immediateTasks.empty()) {
                 _impl->idleNotificationCallback(_impl->id);
 
                 // Wait until new tasks come.
-                _impl->hasTaskEvent.wait(lock, [this]() { return !_impl->tasks.empty(); });
+                _impl->hasTaskEvent.wait(
+                    lock, 
+                    [this]() { return !(_impl->tasks.empty() && _impl->immediateTasks.empty()); });
             }
 
```

Comment:
Add comments describing the new task draining behavior

Reply:
done!